File: packages/types/src/issue-comment-summary.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

/**
 * Category 9, feature 4 - "AI thread summary"
 * (docs/feature-specs/09-ai-features.md in plane-selfhost). Mirrors
 * `IssueCommentSummarySerializer`
 * (apps/api/plane/app/serializers/issue_comment_summary.py). Note
 * `error_message` is deliberately never a field here - the backend never
 * serializes it back to the client (exigence 9); a `FAILED` status is the
 * only signal the UI ever gets.
 */
export type TIssueCommentSummaryStatus = "PENDING" | "COMPLETED" | "FAILED";

export type TIssueCommentSummaryCitation = {
  marker: number;
  comment_id: string;
  snippet: string;
};

export type TIssueCommentSummary = {
  id: string;
  workspace: string;
  project: string;
  issue: string;
  summary_text: string;
  citations: TIssueCommentSummaryCitation[];
  source_comment_count: number;
  status: TIssueCommentSummaryStatus;
  model_used: string;
  generated_by: string | null;
  generated_at: string | null;
  /** True if comments were added/edited/removed on the issue since this
   * summary was generated - computed server-side from `source_comments_hash`. */
  is_stale: boolean;
  created_at: string;
  updated_at: string;
};

File: apps/api/plane/app/serializers/issue_comment_summary.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""
Serializer for `IssueCommentSummary` (category 9, feature 4 - AI thread
summary, docs/feature-specs/09-ai-features.md in plane-selfhost).

`error_message` is intentionally never a field here - exigence 9 requires
the raw provider error to stay server-side only (logged, not shown). The
client only ever learns `status == "FAILED"`, which is enough to render a
generic "couldn't generate the summary, try again" message.

`is_stale` is a computed field (exigence 6): compares the live comment set's
fingerprint against the stored `source_comments_hash` so the frontend can
show a "new comments since this summary" badge without having to reload and
diff the whole comment list itself.
"""

from rest_framework import serializers

from plane.db.models import IssueCommentSummary
from plane.utils.issue_comment_summary import current_comments_hash_for_issue

from .base import BaseSerializer


class IssueCommentSummarySerializer(BaseSerializer):
    is_stale = serializers.SerializerMethodField()

    class Meta:
        model = IssueCommentSummary
        fields = [
            "id",
            "workspace",
            "project",
            "issue",
            "summary_text",
            "citations",
            "source_comment_count",
            "status",
            "model_used",
            "generated_by",
            "generated_at",
            "is_stale",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields

    def get_is_stale(self, obj):
        return current_comments_hash_for_issue(obj.issue_id) != obj.source_comments_hash
